sql/core/src/main/java/org/apache/spark/sql/execution/datasources/orc/JavaOrcColumnarBatchReader.java

```diff
@@ -0,0 +1,503 @@
+
+  /**
+   * Record reader from row batch.
+   */
+  private org.apache.orc.RecordReader recordReader;
+
+  /**
+   * Required Schema.
+   */
+  private StructType requiredSchema;
+
+  /**
+   * ColumnarBatch for vectorized execution by whole-stage codegen.
+   */
+  private ColumnarBatch columnarBatch;
+
+  /**
+   * Writable column vectors of ColumnarBatch.
+   */
+  private WritableColumnVector[] columnVectors;
+
+  /**
+   * The number of rows read and considered to be returned.
+   */
+  private long rowsReturned = 0L;
+
+  /**
+   * Total number of rows.
+   */
+  private long totalRowCount = 0L;
+
+  @Override
+  public Void getCurrentKey() throws IOException, InterruptedException {
+    return null;
+  }
+
+  @Override
+  public ColumnarBatch getCurrentValue() throws IOException, InterruptedException {
+    return columnarBatch;
+  }
+
+  @Override
+  public float getProgress() throws IOException, InterruptedException {
+    return (float) rowsReturned / totalRowCount;
+  }
+
+  @Override
+  public boolean nextKeyValue() throws IOException, InterruptedException {
+    return nextBatch();
+  }
+
+  @Override
+  public void close() throws IOException {
+    if (columnarBatch != null) {
+      columnarBatch.close();
+      columnarBatch = null;
+    }
+    if (recordReader != null) {
+      recordReader.close();
+      recordReader = null;
+    }
+  }
+
+  /**
+   * Initialize ORC file reader and batch record reader.
+   * Please note that `setRequiredSchema` is needed to be called after this.
+   */
+  @Override
+  public void initialize(InputSplit inputSplit, TaskAttemptContext taskAttemptContext)
+      throws IOException, InterruptedException {
+    FileSplit fileSplit = (FileSplit)inputSplit;
+    Configuration conf = taskAttemptContext.getConfiguration();
+    reader = OrcFile.createReader(
+      fileSplit.getPath(),
+      OrcFile.readerOptions(conf)
+        .maxLength(OrcConf.MAX_FILE_LENGTH.getLong(conf))
+        .filesystem(fileSplit.getPath().getFileSystem(conf)));
+
+    Reader.Options options =
+            OrcInputFormat.buildOptions(conf, reader, fileSplit.getStart(), fileSplit.getLength());
+    recordReader = reader.rows(options);
+    totalRowCount = reader.getNumberOfRows();
+  }
+
+  /**
+   * Set required schema and partition information.
+   * With this information, this creates ColumnarBatch with the full schema.
+   */
+  public void setRequiredSchema(
+    TypeDescription orcSchema,
+    int[] requestedColIds,
+    StructType requiredSchema,
+    StructType partitionSchema,
+    InternalRow partitionValues) {
+    batch = orcSchema.createRowBatch(DEFAULT_SIZE);
+    assert(!batch.selectedInUse); // `selectedInUse` should be initialized with `false`.
+
+    StructType resultSchema = new StructType(requiredSchema.fields());
+    for (StructField f : partitionSchema.fields())
+      resultSchema = resultSchema.add(f);
+    this.requiredSchema = requiredSchema;
+    this.requestedColIds = requestedColIds;
```

Comment:
add an assert that `requiredSchema.length == requesredColIds.length`

Reply:
Yep.